internal/v1/handler_blueprints.go

```diff
@@ -252,6 +255,225 @@ func WithRedactedFiles(paths []string) BlueprintBodyOption {
+			var mode *os.FileMode
+			if dir.Mode != nil {
+				// Parse octal mode string to os.FileMode
+				if modeVal, err := strconv.ParseUint(*dir.Mode, 8, 32); err == nil {
+					m := os.FileMode(modeVal)
+					mode = &m
+				}
+			}
+
+			var user interface{}
+			if dir.User != nil {
+				// Handle union type - could be string or int64
+				if userStr, err := dir.User.AsDirectoryUser0(); err == nil {
+					user = userStr
+				} else if userInt, err := dir.User.AsDirectoryUser1(); err == nil {
+					user = userInt
+				}
+			}
+
+			var group interface{}
+			if dir.Group != nil {
+				// Handle union type - could be string or int64
+				if groupStr, err := dir.Group.AsDirectoryGroup0(); err == nil {
+					group = groupStr
+				} else if groupInt, err := dir.Group.AsDirectoryGroup1(); err == nil {
+					group = groupInt
+				}
+			}
+
+			ensureParents := false
+			if dir.EnsureParents != nil {
+				ensureParents = *dir.EnsureParents
+			}
+
+			// Use fsnode.NewDirectory for validation - this handles all path, mode, user, group validation
+			_, err := fsnode.NewDirectory(dir.Path, mode, user, group, ensureParents)
+			if err != nil {
+				return ValidationError{
+					HTTPErrorList: HTTPErrorList{
+						Errors: []HTTPError{{
+							Title:  "Invalid directory customization",
+							Detail: fmt.Sprintf("directory %q: %s", dir.Path, err.Error()),
+						}},
+					},
+				}
+			}
+		}
+	}
+
+	// Validate Filesystem using basic path validation patterns from fsnode library
+	if blueprintRequest.Customizations.Filesystem != nil {
+		for _, fs := range *blueprintRequest.Customizations.Filesystem {
+			// Use the same path validation logic as fsnode (following library patterns)
+			if fs.Mountpoint == "" {
+				return ValidationError{
+					HTTPErrorList: HTTPErrorList{
+						Errors: []HTTPError{{
+							Title:  "Invalid filesystem customization",
+							Detail: "mountpoint must not be empty",
+						}},
+					},
+				}
+			}
+			if fs.Mountpoint[0] != '/' {
+				return ValidationError{
+					HTTPErrorList: HTTPErrorList{
+						Errors: []HTTPError{{
+							Title:  "Invalid filesystem customization",
+							Detail: fmt.Sprintf("mountpoint %q must be absolute", fs.Mountpoint),
+						}},
+					},
+				}
+			}
+			if fs.Mountpoint != filepath.Clean(fs.Mountpoint) {
+				return ValidationError{
+					HTTPErrorList: HTTPErrorList{
+						Errors: []HTTPError{{
+							Title:  "Invalid filesystem customization",
+							Detail: fmt.Sprintf("mountpoint %q must be canonical", fs.Mountpoint),
+						}},
+					},
+				}
+			}
+
+			// Validate minimum size is reasonable
+			if fs.MinSize > 0 && fs.MinSize < 1024*1024 { // 1MB minimum
+				return ValidationError{
+					HTTPErrorList: HTTPErrorList{
+						Errors: []HTTPError{{
+							Title:  "Invalid filesystem customization",
+							Detail: fmt.Sprintf("mountpoint %q minimum size must be at least 1MB", fs.Mountpoint),
+						}},
+					},
+				}
+			}
+		}
+	}
+
+	return nil
+}
```

Comment:
Wow this is a big function, I agree with @avitova in fact I would consider defining a slice of a function type `Validate(ctx, request) *HTTPError or []HTTPError` so all these functions can be executed in a loop. I think this is called Chain or responsibility.

They could even live in a separate package because once we upgrade to UBP these validations could be easily moved out of CRC into blueprint-schema as long as they do not depend on anything major.